runtime/src/bridge.rs

```diff
@@ -0,0 +1,477 @@
+use frame_support::traits::{Currency, ExistenceRequirement::AllowDeath, Get};
+use frame_support::{decl_error, decl_event, decl_module, dispatch::DispatchResult, ensure};
+use frame_system::{self as system, ensure_signed};
+use sp_runtime::traits::EnsureOrigin;
+use sp_std::prelude::*;
+
+type ResourceId = chainbridge::ResourceId;
+
+pub trait Trait: system::Trait + chainbridge::Trait {
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+    /// Specifies the origin check provided by the chainbridge for calls that can only be called by the chainbridge pallet
+    type BridgeOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
+
+    /// Ids can be defined by the runtime and passed in, perhaps from blake2b_128 hashes.
+    type HashId: Get<ResourceId>;
+    type NativeTokenId: Get<ResourceId>;
+}
+
+decl_event! {
+    pub enum Event<T> where
+        <T as frame_system::Trait>::Hash,
+    {
+        Remark(Hash),
+    }
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait>{
+        InvalidTransfer,
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        const HashId: ResourceId = T::HashId::get();
+        const NativeTokenId: ResourceId = T::NativeTokenId::get();
+
+        fn deposit_event() = default;
+
+        //
+        // Initiation calls. These start a chainbridge transfer.
+        //
+
+        /// Transfers an arbitrary hash to a (whitelisted) destination chain.
+        pub fn transfer_hash(origin, hash: T::Hash, dest_id: chainbridge::ChainId) -> DispatchResult {
+            ensure_signed(origin)?;
+
+            let resource_id = T::HashId::get();
+            let metadata: Vec<u8> = hash.as_ref().to_vec();
+            <chainbridge::Module<T>>::transfer_generic(dest_id, resource_id, metadata)
+        }
+
+        /// Transfers some amount of the native token to some recipient on a (whitelisted) destination chain.
+        pub fn transfer_native(origin, amount: u32, recipient: Vec<u8>, dest_id: chainbridge::ChainId) -> DispatchResult {
```

Comment:
How are the weights/fees defined here? I think we need to define them here.

Reply:
I have no idea how weights/fees would be here.

Reply:
@philipstanislaus this is the entrypoint for a user in the centchain to move tokens to ethereum. Do you have an idea of what weight amount should be set? Prob a bit more expensive than a regular balance transfer?

Reply:
shouldn't it be `amount: u64`?

Reply:
we'll have to discuss weights, @mikiquantum I don't think @philipstanislaus did any calculations but let's fix the weights at a later point.

Reply:
Technically a Balance in substrate is u128. The bridge pallet only supports u32 at the moment. That would be a problem if the bridge uses minimum units for any chain interaction.
> we'll have to discuss weights, @mikiquantum I don't think @philipstanislaus did any calculations but let's fix the weights at a later point.
Can we add at least now the same as any balance transfer weight?

Reply:
> shouldn't it be amount: u64

I think this should be using Balance for the type.

Reply:
https://github.com/ChainSafe/chainbridge-substrate/issues/44